// @forgerock/javascript-sdk/dist/fr-auth/callbacks/attribute-input-callback.js
import r from "./index.js";
class s extends r {
  /**
   * @param payload The raw payload returned by OpenAM
   */
  constructor(e) {
    super(e), this.payload = e;
  }
  /**
   * Gets the attribute name.
   */
  getName() {
    return this.getOutputByName("name", "");
  }
  /**
   * Gets the attribute prompt.
   */
  getPrompt() {
    return this.getOutputByName("prompt", "");
  }
  /**
   * Gets whether the attribute is required.
   */
  isRequired() {
    return this.getOutputByName("required", !1);
  }
  /**
   * Gets the callback's failed policies.
   */
  getFailedPolicies() {
    const e = this.getOutputByName(
      "failedPolicies",
      []
    );
    try {
      return e.map((t) => JSON.parse(t));
    } catch {
      throw new Error(
        'Unable to parse "failed policies" from the ForgeRock server. The JSON within `AttributeInputCallback` was either malformed or missing.'
      );
    }
  }
  /**
   * Gets the callback's applicable policies.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  getPolicies() {
    return this.getOutputByName("policies", {});
  }
  /**
   * Set if validating value only.
   */
  setValidateOnly(e) {
    this.setInputValue(e, /validateOnly/);
  }
  /**
   * Sets the attribute's value.
   */
  setValue(e) {
    this.setInputValue(e);
  }
}
export {
  s as default
};
