// @forgerock/javascript-sdk/dist/fr-auth/callbacks/kba-create-callback.js
import n from "./index.js";
class u extends n {
  /**
   * @param payload The raw payload returned by OpenAM
   */
  constructor(e) {
    super(e), this.payload = e;
  }
  /**
   * Gets the callback prompt.
   */
  getPrompt() {
    return this.getOutputByName("prompt", "");
  }
  /**
   * Gets the callback's list of pre-defined security questions.
   */
  getPredefinedQuestions() {
    return this.getOutputByName("predefinedQuestions", []);
  }
  /**
   * Gets whether the user can define questions.
   */
  isAllowedUserDefinedQuestions() {
    return this.getOutputByName("allowUserDefinedQuestions", !1);
  }
  /**
   * Sets the callback's security question.
   */
  setQuestion(e) {
    this.setValue("question", e);
  }
  /**
   * Sets the callback's security question answer.
   */
  setAnswer(e) {
    this.setValue("answer", e);
  }
  setValue(e, s) {
    if (!this.payload.input)
      throw new Error("KBA payload is missing input");
    const t = this.payload.input.find((i) => i.name.endsWith(e));
    if (!t)
      throw new Error(`No input has name ending in "${e}"`);
    t.value = s;
  }
}
export {
  u as default
};
