// @forgerock/javascript-sdk/dist/fr-auth/callbacks/ping-protect-initialize-callback.js
import s from "./index.js";
class l extends s {
  /**
   * @param payload The raw payload returned by OpenAM
   */
  constructor(t) {
    super(t), this.payload = t;
  }
  /**
   * Get callback's initialization config settings
   */
  getConfig() {
    const t = this.getOutputByName(
      "signalsInitializationOptions",
      void 0
    );
    if (t !== null && typeof t == "object" && !Array.isArray(t))
      return t;
    const e = this.getOutputByName(
      "agentIdentification",
      void 0
    ), i = this.getOutputByName("agentTimeout", void 0), a = this.getOutputByName("agentPort", void 0);
    return {
      // Required parameter
      envId: this.getOutputByName("envId", ""),
      // Optional parameters
      ...e !== void 0 ? { agentIdentification: e } : {},
      ...i !== void 0 ? { agentTimeout: i } : {},
      ...a !== void 0 ? { agentPort: a } : {},
      behavioralDataCollection: this.getOutputByName("behavioralDataCollection", !0),
      disableTags: this.getOutputByName("disableTags", !1),
      universalDeviceIdentification: this.getOutputByName(
        "universalDeviceIdentification",
        !1
      ),
      // Deprecated parameters
      consoleLogEnabled: this.getOutputByName("consoleLogEnabled", !1),
      deviceAttributesToIgnore: this.getOutputByName("deviceAttributesToIgnore", []),
      customHost: this.getOutputByName("customHost", ""),
      lazyMetadata: this.getOutputByName("lazyMetadata", !1),
      deviceKeyRsyncIntervals: this.getOutputByName("deviceKeyRsyncIntervals", 14),
      enableTrust: this.getOutputByName("enableTrust", !1),
      disableHub: this.getOutputByName("disableHub", !1)
    };
  }
  setClientError(t) {
    this.setInputValue(t, /clientError/);
  }
}
export {
  l as default
};
