// @forgerock/javascript-sdk/dist/fr-webauthn/helpers.js
import { WebAuthnOutcomeType as o } from "./enums.js";
function s(r) {
  return r || [];
}
function f(r) {
  const n = new Uint8Array(r);
  return new TextDecoder().decode(n);
}
function u(r) {
  return r ? r[1] : "";
}
function p(r) {
  if (!r || r === "" || r === "[]")
    return [];
  try {
    return r.split("}").filter((e) => !!e && e !== "]").map((e) => {
      const t = c(e);
      return {
        id: new Int8Array(t).buffer,
        type: "public-key"
      };
    });
  } catch {
    const e = new Error("Transforming credential object to string failed");
    throw e.name = o.EncodingError, e;
  }
}
function c(r) {
  const n = /new Int8Array\((.+)\)/.exec(r);
  return n === null || n.length < 2 ? [] : JSON.parse(n[1]);
}
function d(r) {
  if (Array.isArray(r))
    return r;
  if (typeof r == "string")
    return r && r[0] === "[" ? JSON.parse(r) : (r = r.replace(/(\w+):/g, '"$1":'), JSON.parse(`[${r}]`));
}
function a(r) {
  return r.includes("rpId") ? r.replace(/rpId: "(.+)",/, "$1") : r.replace(/id: "(.+)",/, "$1");
}
export {
  f as arrayBufferToString,
  s as ensureArray,
  u as getIndexOne,
  p as parseCredentials,
  c as parseNumberArray,
  d as parsePubKeyArray,
  a as parseRelyingPartyId
};
