// @forgerock/javascript-sdk/dist/fr-webauthn/index.js
import { CallbackType as y } from "../auth/enums.js";
import { FRLogger as k } from "../util/logger.js";
import { WebAuthnStepType as f, WebAuthnOutcomeType as l, WebAuthnOutcome as m } from "./enums.js";
import { arrayBufferToString as C, parseCredentials as w, parseRelyingPartyId as A, parsePubKeyArray as E } from "./helpers.js";
import { parseWebAuthnAuthenticateText as S, parseWebAuthnRegisterText as P } from "./script-parser.js";
import { withTimeout as I } from "../util/timeout.js";
const $ = 2e3;
class W {
  /**
   * Determines if the given step is a WebAuthn step.
   *
   * @param step The step to evaluate
   * @return A WebAuthnStepType value
   */
  static getWebAuthnStepType(e) {
    const a = this.getOutcomeCallback(e), t = this.getMetadataCallback(e), n = this.getTextOutputCallback(e);
    return a && t ? t.getOutputValue("data")?.pubKeyCredParams ? f.Registration : f.Authentication : a && n ? n.getMessage().includes("pubKeyCredParams") ? f.Registration : f.Authentication : f.None;
  }
  /**
   * Checks if the browser supports WebAuthn.
   *
   * @return boolean indicating if WebAuthn is available
   */
  static isWebAuthnSupported() {
    return !!window.PublicKeyCredential;
  }
  /**
   * Checks if the browser supports conditional mediation (autofill) for WebAuthn.
   *
   * @return Promise<boolean> indicating if conditional mediation is available
   */
  static async isConditionalMediationSupported() {
    if (!this.isWebAuthnSupported())
      return !1;
    try {
      return I(PublicKeyCredential.isConditionalMediationAvailable(), $);
    } catch {
      k.warn("Conditional mediation check timed out");
    }
    return !1;
  }
  /**
   * Populates the step with the necessary authentication outcome.
   * Automatically handles conditional mediation if indicated by the server metadata.
   *
   * @param step The step that contains WebAuthn authentication data
   * @param optionsTransformer Augments the derived options with custom behaviour
   * @return The populated step
   */
  static async authenticate(e, a = (t) => t) {
    const { hiddenCallback: t, metadataCallback: n, textOutputCallback: i } = this.getCallbacks(e);
    if (t && (n || i)) {
      const r = {};
      try {
        if (n) {
          const o = n.getOutputValue("data"), u = o.mediation;
          if (u === "conditional" && !await this.isConditionalMediationSupported()) {
            const p = new Error(
              "Conditional mediation was requested, but is not supported by this browser."
            );
            throw p.name = l.NotSupportedError, p;
          }
          r.publicKey = this.createAuthenticationPublicKey(o), r.mediation = u;
        } else if (i) {
          const o = this.extractMetadata(i.getMessage());
          o ? r.publicKey = this.createAuthenticationPublicKey(
            o
          ) : r.publicKey = S(i.getMessage());
        } else
          throw new Error("No Credential found from Public Key");
      } catch (o) {
        throw o instanceof Error ? o.name === l.NotSupportedError ? (t.setInputValue(m.Unsupported), o) : (t.setInputValue(`${m.Error}::${o.name}:${o.message}`), o) : o;
      }
      const s = await this.getAuthenticationCredential(
        a(r)
      ), c = this.getAuthenticationOutcome(s);
      return n && n.getOutputValue("data")?.supportsJsonResponse && s && "authenticatorAttachment" in s ? (t.setInputValue(
        JSON.stringify({
          authenticatorAttachment: s.authenticatorAttachment,
          legacyData: c
        })
      ), e) : (t.setInputValue(c), e);
    } else {
      const r = new Error("Incorrect callbacks for WebAuthn authentication");
      throw r.name = l.DataError, t?.setInputValue(`${m.Error}::${r.name}:${r.message}`), r;
    }
  }
  /**
   * Populates the step with the necessary registration outcome.
   *
   * @param step The step that contains WebAuthn registration data
   * @return The populated step
   */
  // Can make this generic const in Typescript 5.0 > and the name itself will
  // be inferred from the type so `typeof deviceName` will not just return string
  // but the actual name of the deviceName passed in as a generic.
  static async register(e, a) {
    const { hiddenCallback: t, metadataCallback: n, textOutputCallback: i } = this.getCallbacks(e);
    if (t && (n || i)) {
      let r, s = null;
      try {
        let c;
        if (n) {
          const o = n.getOutputValue("data");
          c = this.createRegistrationPublicKey(o), s = await this.getRegistrationCredential(
            c
          ), r = this.getRegistrationOutcome(s);
        } else if (i) {
          const o = this.extractMetadata(i.getMessage());
          o ? c = this.createRegistrationPublicKey(o) : c = P(i.getMessage()), s = await this.getRegistrationCredential(
            c
          ), r = this.getRegistrationOutcome(s);
        } else
          throw new Error("No Credential found from Public Key");
      } catch (c) {
        throw c instanceof Error ? c.name === l.NotSupportedError ? (t.setInputValue(m.Unsupported), c) : (t.setInputValue(`${m.Error}::${c.name}:${c.message}`), c) : c;
      }
      return n && n.getOutputValue("data")?.supportsJsonResponse && s && "authenticatorAttachment" in s ? (t.setInputValue(
        JSON.stringify({
          authenticatorAttachment: s.authenticatorAttachment,
          legacyData: a && a.length > 0 ? `${r}::${a}` : r
        })
      ), e) : (t.setInputValue(
        a && a.length > 0 ? `${r}::${a}` : r
      ), e);
    } else {
      const r = new Error("Incorrect callbacks for WebAuthn registration");
      throw r.name = l.DataError, t?.setInputValue(`${m.Error}::${r.name}:${r.message}`), r;
    }
  }
  /**
   * Returns an object containing the two WebAuthn callbacks.
   *
   * @param step The step that contains WebAuthn callbacks
   * @return The WebAuthn callbacks
   */
  static getCallbacks(e) {
    const a = this.getOutcomeCallback(e), t = this.getMetadataCallback(e), n = this.getTextOutputCallback(e), i = {
      hiddenCallback: a
    };
    return t ? i.metadataCallback = t : n && (i.textOutputCallback = n), i;
  }
  /**
   * Returns the WebAuthn metadata callback containing data to pass to the browser
   * Web Authentication API.
   *
   * @param step The step that contains WebAuthn callbacks
   * @return The metadata callback
   */
  static getMetadataCallback(e) {
    return e.getCallbacksOfType(y.MetadataCallback).find((a) => {
      const t = a.getOutputByName("data", void 0);
      return t && t.hasOwnProperty("relyingPartyId");
    });
  }
  /**
   * Returns the WebAuthn hidden value callback where the outcome should be populated.
   *
   * @param step The step that contains WebAuthn callbacks
   * @return The hidden value callback
   */
  static getOutcomeCallback(e) {
    return e.getCallbacksOfType(y.HiddenValueCallback).find((a) => a.getOutputByName("id", "") === "webAuthnOutcome");
  }
  /**
   * Returns the WebAuthn metadata callback containing data to pass to the browser
   * Web Authentication API.
   *
   * @param step The step that contains WebAuthn callbacks
   * @return The metadata callback
   */
  static getTextOutputCallback(e) {
    return e.getCallbacksOfType(y.TextOutputCallback).find((a) => {
      const t = a.getOutputByName("message", void 0);
      return t && t.includes("webAuthnOutcome");
    });
  }
  /**
   * Retrieves the credential from the browser Web Authentication API.
   *
   * @param options The options associated with the request
   * @return The credential
   */
  static async getAuthenticationCredential(e) {
    if (!this.isWebAuthnSupported()) {
      const t = new Error("PublicKeyCredential not supported by this browser");
      throw t.name = l.NotSupportedError, t;
    }
    return await navigator.credentials.get({
      ...e,
      signal: this.createAbortController().signal
    });
  }
  /**
   * Converts an authentication credential into the outcome expected by OpenAM.
   *
   * @param credential The credential to convert
   * @return The outcome string
   */
  static getAuthenticationOutcome(e) {
    if (e === null) {
      const a = new Error("No credential generated from authentication");
      throw a.name = l.UnknownError, a;
    }
    try {
      const a = C(e.response.clientDataJSON), t = e.response, n = new Int8Array(
        t.authenticatorData
      ).toString(), i = new Int8Array(t.signature).toString(), r = C(e.response.userHandle);
      let s = `${a}::${n}::${i}::${e.id}`;
      return r && (s = `${s}::${r}`), s;
    } catch {
      const t = new Error("Transforming credential object to string failed");
      throw t.name = l.EncodingError, t;
    }
  }
  /**
   * Retrieves the credential from the browser Web Authentication API.
   *
   * @param options The public key options associated with the request
   * @return The credential
   */
  static async getRegistrationCredential(e) {
    if (!this.isWebAuthnSupported()) {
      const t = new Error("PublicKeyCredential not supported by this browser");
      throw t.name = l.NotSupportedError, t;
    }
    return await navigator.credentials.create({
      publicKey: e
    });
  }
  /**
   * Converts a registration credential into the outcome expected by OpenAM.
   *
   * @param credential The credential to convert
   * @return The outcome string
   */
  static getRegistrationOutcome(e) {
    if (e === null) {
      const a = new Error("No credential generated from registration");
      throw a.name = l.UnknownError, a;
    }
    try {
      const a = C(e.response.clientDataJSON), t = e.response, n = new Int8Array(
        t.attestationObject
      ).toString();
      return `${a}::${n}::${e.id}`;
    } catch {
      const t = new Error("Transforming credential object to string failed");
      throw t.name = l.EncodingError, t;
    }
  }
  /**
   * Converts authentication tree metadata into options required by the browser
   * Web Authentication API.
   *
   * @param metadata The metadata provided in the authentication tree MetadataCallback
   * @return The Web Authentication API request options
   */
  static createAuthenticationPublicKey(e) {
    const {
      acceptableCredentials: a,
      allowCredentials: t,
      _allowCredentials: n,
      challenge: i,
      relyingPartyId: r,
      _relyingPartyId: s,
      timeout: c,
      userVerification: o,
      extensions: u
    } = e;
    let d;
    n && Array.isArray(n) ? d = n.map((g) => ({
      ...g,
      id: new Int8Array(g.id)
    })) : d = w(t || a || "");
    const p = s || A(r), h = {
      challenge: Uint8Array.from(atob(i), (g) => g.charCodeAt(0)).buffer,
      timeout: c
    };
    return d && d.length > 0 && (h.allowCredentials = d), o && (h.userVerification = o), p && (h.rpId = p), u && Object.keys(u).length > 0 && (h.extensions = u), h;
  }
  /**
   * Converts authentication tree metadata into options required by the browser
   * Web Authentication API.
   *
   * @param metadata The metadata provided in the authentication tree MetadataCallback
   * @return The Web Authentication API request options
   */
  static createRegistrationPublicKey(e) {
    const { pubKeyCredParams: a } = e, t = E(a);
    if (!t) {
      const b = new Error("Missing pubKeyCredParams property from registration options");
      throw b.name = l.DataError, b;
    }
    const n = w(e.excludeCredentials), {
      attestationPreference: i,
      authenticatorSelection: r,
      challenge: s,
      relyingPartyId: c,
      relyingPartyName: o,
      timeout: u,
      userId: d,
      userName: p,
      displayName: h
    } = e, g = A(c), O = {
      name: o,
      ...g && { id: g }
    };
    return {
      attestation: i,
      authenticatorSelection: JSON.parse(r),
      challenge: Uint8Array.from(atob(s), (b) => b.charCodeAt(0)).buffer,
      ...n.length && { excludeCredentials: n },
      pubKeyCredParams: t,
      rp: O,
      timeout: u,
      user: {
        displayName: h || p,
        id: Int8Array.from(d.split("").map((b) => b.charCodeAt(0))),
        name: h || p
      }
    };
  }
  static createAbortController() {
    window.PingWebAuthnAbortController?.abort();
    const e = new AbortController();
    return window.PingWebAuthnAbortController = e, e;
  }
  static extractMetadata(e) {
    const t = e.match(/^var scriptContext = (.*);*$/m)?.[1];
    return t ? JSON.parse(t) : null;
  }
}
export {
  m as WebAuthnOutcome,
  l as WebAuthnOutcomeType,
  f as WebAuthnStepType,
  W as default
};
